### ISSUE TITLE
[Question] Semi optional generator parameters

### ISSUE THREAD

**Issue by Kakadu** [ocaml-ppx/ppxlib#22]
This is a continuation of https://discuss.ocaml.org/t/semi-optional-deriving-arguments-using-ppxlib/1817

@diml the `default` field and the whole type `'a Deriving.Args.param` is not exposed to the `mli`. So, I can't do what I want without patching `ppxlib`. Am I right?

**Comment by diml** [comment]
Yes, that's right. I thought that this is what you were asking, i.e. how to patch ppxlib?

However, looking again at the code, setting `default` is not actually the right solution as `default` is the value used when the argument is missing.

In the parsetree, `~foo` is represented as `~foo:foo`, and that's how `flag` works. So technically you could achieve what you want without modifying ppxlib, by simply matching against `foo`.

That said, a patch to add a new function in `Args` making this explicit would be welcome. For instance something like this:

```
type 'a opt =
  | Missing
  | Present_without_argument
  | Present_with_argument of 'a

val opt : ... -> 'a opt param
```

**Comment by Kakadu** [comment]
I will try hacky solution using `~foo:foo` first. Don't expect any patch soon

**Comment by Kakadu** [comment]
@diml Also, I think, that using module `Deriving.Arg` we can't specify that generator takes an arbitrary number of arguments of the similar structure. Should I make a separate issue about this?

**Comment by diml** [comment]
Do you mean you want to write something like this: `[@@deriving foo ~x:... ~x:... ~x:...]`, i.e. with the same argument several times?

**Comment by Kakadu** [comment]
@diml I meant `[@@deriving foo ~name1:... ~name2:... ~name3:.... ....... ~name100500:....]` where all `name%d` should be arbitrary. As you have probably already guessed I want my generator (`foo` here) be extendable by plugins.

**Comment by diml** [comment]
I see. Yh, please open a separate issue about this
